### PR TITLE
Fix for issue #117 #114 #113 #88

### PR DESCRIPTION
**Pull request by NiteshKant**
Issue #117: This was a major issue with the RxClient implementation which was mutating (updating ChannelInitializer) in the bootstrap for every connect. The reason for doing this was to have a ClientLifecycleListener provide connection events. This design was naively taken from the RxServer but does not apply for the client as the LifecycleListener requires an observer which changes for the client per connection.
This change fixes this issue by removing the newConnection notification responsibility from the LifecycleListener and have it done by the ClientChannelFactory. This also enabled to remove a few other redundant constructs like the AbstractChannelFactory which was unnecessarily complex. So, the code looks much cleaner now :)

Issue #114: rx-netty core was using numerus for LongAdder implementation. Now it uses netty's internal backport of LongAdder.

Issue #113: Rewrote the Redirect implementation as an Rx operator. This makes the code much cleaner and provides a very useful feature of being able to elect which HttpClient implementation must be used for executing redirects. This was a major limitation of the previous approach.

Issue #88: Provide an implementation of CompositeHttpClient which is a composite of HttpClients per ServerInfo. This provides additional submit() methods which also takes a host, port. By default this client submits requests to localhost.
RxNetty now has two flavors of createHttpRequest() that takes a HttpClientRequest with the full URI set and uses an appropriate HttpClient instance to execute this request.
Now, one can use a shorthand:

```
RxNetty.createHttpRequest(HttpClientRequest.createGet("http://localhost:8080/hello"))
```

to submit a request with full URI which was not previously possible.

### CODE REVIEW

File: rx-netty/src/main/java/io/reactivex/netty/protocol/http/client/DefaultRedirectHandler.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    }
+
+    protected Observable<HttpClientResponse<O>> redirect(HttpClientRequest<I> redirectRequest) {
+        return client.submit(redirectRequest, config);
+    }
+
+    @Override
+    public boolean requiresRedirect(RedirectionContext context, HttpClientResponse<O> response) {
+        int statusCode = response.getStatus().code();
+        return statusCode >= 301 && statusCode <= 308;
```
Comment (allenxwang) [inline rx-netty/src/main/java/io/reactivex/netty/protocol/http/client/DefaultRedirectHandler.java]: 304 is for HTTP caching, not for redirect.

**Comment by benjchristensen** [comment]
How about for the most common cases we have overloads like this?

``` java
RxNetty.createHttpGet("http://localhost:8080/hello")
```

File: rx-netty/src/main/java/io/reactivex/netty/protocol/http/client/DefaultRedirectHandler.java (hunk 1 above)
Comment (NiteshKant) [inline rx-netty/src/main/java/io/reactivex/netty/protocol/http/client/DefaultRedirectHandler.java]: Thanks! I updated the code to only redirect for 301,302,303, 307, 308.

**Comment by NiteshKant** [comment]
@benjchristensen yes I will add that method. The caveat being that if the user wants to add any header, the HttpClientRequest builder has to be used.
I will also add similar method for DELETE.
Shorthands for PUT & POST becomes a bit more complex as we have a few ways of specifying the content. So, I will stay away from them as of now.

I personally feel we should give a bit more thought on making this more usable by providing a more fluent DSL more on the lines you suggested offline i.e. by having an extension to Observable. I will put together my thoughts on that and open a different issue.

**Pull request merged**